pkg/fqdn/re/re.go

```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+// Package re provides a simple function to access compile regex objects for
+// the FQDN subsystem.
+package re
+
+import (
+	"fmt"
+	"regexp"
+
+	lru "github.com/golang/groupcache/lru"
+
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "fqdn/re")
+)
+
+// CompileRegex compiles a pattern p into a regex and returns the regex object.
+// The regex object will be cached by an LRU. If p has already been compiled
+// and cached, this function will return the cached regex object. If not
+// already cached, it will compile p into a regex object and cache it in the
+// LRU.
+func (c *RegexCompileLRU) CompileRegex(p string) (*regexp.Regexp, error) {
+	c.Lock()
+	r, ok := c.Get(p)
+	c.Unlock()
+	if ok {
+		return r.(*regexp.Regexp), nil
+	}
+	n, err := regexp.Compile(p)
+	if err != nil {
+		return nil, fmt.Errorf("failed to compile regex: %w", err)
+	}
+	c.Lock()
+	c.Add(p, n)
+	c.Unlock()
+	return n, nil
+}
+
+// InitRegexCompileLRU initializes the singleton LRU cache.
+func InitRegexCompileLRU(size int) (*RegexCompileLRU, error) {
```

Comment:
nit, but typically when Cilium code creates an object it will begin with `New...()`. We can touch this up before merge though.

```suggestion
// NewRegexCompileLRU initializes the singleton LRU cache.
func NewRegexCompileLRU(size int) (*RegexCompileLRU, error) {
```